### PR TITLE
[yhirose/cpp-httplib] Add to oss-fuzz 

### PR DESCRIPTION
**Pull request by omjego**
- Depends on https://github.com/yhirose/cpp-httplib/pull/684
- Will change repo link in Dockerfile after above gets merged.

Testing locally: 
``` 
cd oss-fuzz
export PROJECT_NAME=cpp-httplib  
python infra/helper.py build_fuzzers --sanitizer address $PROJECT_NAME
python infra/helper.py run_fuzzer $PROJECT_NAME server_fuzzer
```

### CODE REVIEW

**Comment by inferno-chromium** [comment]
Check out memory sanitizer and address i386 build breakages. You can also disable these configs in first iteration.

**Comment by omjego** [comment]
Hey @inferno-chromium 
 So I was trying to debug fixing the build breakages, some of the issue that I found are
1. The dictionary we were using had  a "\t" character which was messing up with afl engine. 
  ```[-] PROGRAM ABORT : Non-printable characters in line 107.  Location : load_extras_file(), afl-fuzz.c:1621```
   I've removed the char, but not sure why its not supported. 
2. msan + libfuzzer: The cpp-httplib library uses openssl1.1.1 (installed via Dockerfile) and msan (from `gcr.io/oss-fuzz-base/msan-libs-builder`) brings in 1.0.0, which was getting statically linked and was giving linking errors. So I tried linking openssl1.1.1 libs using specific path. 
 Though while doing build check for above I got following memory leak 
 ```
BAD BUILD: /tmp/not-out/server_fuzzer seems to have either startup crash or exit:
/tmp/not-out/server_fuzzer -rss_limit_mb=2560 -timeout=25 -seed=1337 -runs=4 -dict=server_fuzzer.dict < /dev/null
Uninitialized bytes in __interceptor_fopen64 at offset 0 inside [0x702000016a00, 27)
==79==WARNING: MemorySanitizer: use-of-uninitialized-value
    #0 0xac51d8 in openssl_fopen /src/openssl-1.1.1f/crypto/o_fopen.c:114:12
    #1 0xa7c3f8 in BIO_new_file /src/openssl-1.1.1f/crypto/bio/bss_file.c:62:18
    #2 0xbbfda1 in def_load /src/openssl-1.1.1f/crypto/conf/conf_def.c:164:10
    #3 0xb5416e in NCONF_load /src/openssl-1.1.1f/crypto/conf/conf_lib.c:214:12
    #4 0xa86192 in CONF_modules_load_file /src/openssl-1.1.1f/crypto/conf/conf_mod.c:129:9
    #5 0xb54475 in openssl_config_int /src/openssl-1.1.1f/crypto/conf/conf_sap.c:69:11
    #6 0xabb43f in ossl_init_config /src/openssl-1.1.1f/crypto/init.c:293:15
    #7 0xabb1d8 in ossl_init_config_ossl_ /src/openssl-1.1.1f/crypto/init.c:291:1
    #8 0x7f8039c9aa98 in __pthread_once_slow (/lib/x86_64-linux-gnu/libpthread.so.0+0xea98)
    #9 0xb02db8 in CRYPTO_THREAD_run_once /src/openssl-1.1.1f/crypto/threads_pthread.c:118:9
    #10 0xabae40 in OPENSSL_init_crypto /src/openssl-1.1.1f/crypto/init.c:701:15
    #11 0xa28cf7 in OPENSSL_init_ssl /src/openssl-1.1.1f/ssl/ssl_init.c:205:10
    #12 0x5291e6 in httplib::detail::SSLInit::SSLInit() /src/cpp-httplib/test/fuzzing/../../httplib.h:5618:5
    #13 0x41fcba in __cxx_global_var_init /src/cpp-httplib/test/fuzzing/../../httplib.h:5691:16
    #14 0x41fe24 in _GLOBAL__sub_I_server_fuzzer.cc /src/cpp-httplib/test/fuzzing/server_fuzzer.cc
    #15 0xd406bc in __libc_csu_init (/tmp/not-out/server_fuzzer+0xd406bc)
    #16 0x7f80394c47ce in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x207ce)
    #17 0x421e78 in _start (/tmp/not-out/server_fuzzer+0x421e78)

DEDUP_TOKEN: openssl_fopen--BIO_new_file--def_load
  Uninitialized value was created by a heap allocation
    #0 0x4d4dad in malloc /src/llvm-project/compiler-rt/lib/msan/msan_interceptors.cpp:901:3
    #1 0xabd5e6 in CRYPTO_malloc /src/openssl-1.1.1f/crypto/mem.c:222:11
    #2 0xa86267 in CONF_get1_default_config_file /src/openssl-1.1.1f/crypto/conf/conf_mod.c:496:12
    #3 0xa8617b in CONF_modules_load_file /src/openssl-1.1.1f/crypto/conf/conf_mod.c:123:16
    #4 0xb54475 in openssl_config_int /src/openssl-1.1.1f/crypto/conf/conf_sap.c:69:11
    #5 0xabb43f in ossl_init_config /src/openssl-1.1.1f/crypto/init.c:293:15
    #6 0xabb1d8 in ossl_init_config_ossl_ /src/openssl-1.1.1f/crypto/init.c:291:1
    #7 0x7f8039c9aa98 in __pthread_once_slow (/lib/x86_64-linux-gnu/libpthread.so.0+0xea98)

DEDUP_TOKEN: malloc--CRYPTO_malloc--CONF_get1_default_config_file
SUMMARY: MemorySanitizer: use-of-uninitialized-value /src/openssl-1.1.1f/crypto/o_fopen.c:114:12 in openssl_fopen
Exiting
ERROR: 100% of fuzz targets seem to be broken. See the list above for a detailed information.

```

Which seems like a false positive (If it is a false +ve, is it cause openssl isn't built with [msan](https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo#introduction)?) Not sure whats the issue. 

For now, as you suggested, we'll remove msan in the first iteration.

**Comment by omjego** [comment]
@yhirose Can you please review this PR as well, especially [project.yaml](https://github.com/google/oss-fuzz/pull/4508/files#diff-18d8ed2d457c6a575939aeb5b290245d) 
Refer: https://google.github.io/oss-fuzz/getting-started/new-project-guide/#primary

File: projects/cpp-httplib/project.yaml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+homepage: "https://github.com/yhirose/cpp-httplib"
+language: c++
+primary_contact: "yuji.hirose.bug@gmail.com"
+auto_ccs :
+  - "omkarjadhav@google.com"
+sanitizers:
+  - address
+  - dataflow
+  - undefined
+architectures:
```
Comment (inferno-chromium) [inline projects/cpp-httplib/project.yaml:10]: This field is not needed, x86_64 is by default.

File: projects/cpp-httplib/build.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+# build fuzz targets specified  in test/Makefile
```
Comment (inferno-chromium) [inline projects/cpp-httplib/build.sh:18]: "# Build fuzz targets specified in test/Makefile."

File: projects/cpp-httplib/build.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+cd test/fuzzing && make -j$(nproc) server_fuzzer
+
+# Copy the fuzzer executables, zip-ed corpora, option and dictionary files to $OUT
+find . -name '*_fuzzer' -exec cp -v '{}' $OUT ';'          # Copy fuzz-target.
+find . -name '*_fuzzer.dict' -exec cp -v '{}' $OUT ';'     # Copy dictionaries.
+find . -name '*_fuzzer_seed_corpus.zip' -exec cp -v '{}' $OUT ';' # Copy seed corpora
```
Comment (inferno-chromium) [inline projects/cpp-httplib/build.sh:24]: Add missing newline.

File: projects/cpp-httplib/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+# Install openssl 1.1.1
+RUN curl -o openssl-1.1.1f.tar.gz https://www.openssl.org/source/openssl-1.1.1f.tar.gz && \
+    tar xzvf  openssl-1.1.1f.tar.gz
+RUN  cd openssl-1.1.1f && ./config && make && make install_sw
+
+# TODO use remote repo instead of forked one.
+RUN git clone -b http-oss-fuzz https://github.com/omjego/cpp-httplib.git cpp-httplib
+
+WORKDIR cpp-httplib
+COPY build.sh $SRC/
```
Comment (inferno-chromium) [inline projects/cpp-httplib/Dockerfile:35]: Mising newline.

File: projects/cpp-httplib/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+#Install zlib
+RUN apt-get update && apt-get install -y zlib1g-dev
+
+# Install openssl 1.1.1
+RUN curl -o openssl-1.1.1f.tar.gz https://www.openssl.org/source/openssl-1.1.1f.tar.gz && \
+    tar xzvf  openssl-1.1.1f.tar.gz
+RUN  cd openssl-1.1.1f && ./config && make && make install_sw
+
+# TODO use remote repo instead of forked one.
```
Comment (inferno-chromium) [inline projects/cpp-httplib/Dockerfile:31]: Yes use official repo.

File: projects/cpp-httplib/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+FROM gcr.io/oss-fuzz-base/base-builder
+RUN apt-get update && apt-get install -y make autoconf automake libtool
+
+#Install brotli
```
Comment (inferno-chromium) [inline projects/cpp-httplib/Dockerfile:20]: "# Install brotli."
Capitalize comments, end with period.

File: projects/cpp-httplib/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+RUN apt-get install -y libbrotli-dev
+
+#Install zlib
+RUN apt-get update && apt-get install -y zlib1g-dev
+
+# Install openssl 1.1.1
+RUN curl -o openssl-1.1.1f.tar.gz https://www.openssl.org/source/openssl-1.1.1f.tar.gz && \
+    tar xzvf  openssl-1.1.1f.tar.gz
+RUN  cd openssl-1.1.1f && ./config && make && make install_sw
```
Comment (inferno-chromium) [inline projects/cpp-httplib/Dockerfile:29]: "RUN cd"
remove extra space.

**Review by inferno-chromium — APPROVED** [review]

**Comment by omjego** [comment]
Hey @inferno-chromium the [PR](https://github.com/yhirose/cpp-httplib/pull/684) in official repo is merged. And I've done the remaining changes. This can be merged now or let me know if there are any changes. :)

**Review by inferno-chromium — APPROVED** [review]

**Pull request merged**